File: app/src/main/java/com/cn29/aac/datasource/api/ApiResponse.java
package com.cn29.aac.datasource.api;

/**
 * Created by Charles Ng on 3/10/2017.
 */

import java.io.IOException;
import java.util.Collections;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.collection.ArrayMap;
import retrofit2.Response;

/**
 * Common class used by API responses.
 */
public class ApiResponse<T> {

  private static final Pattern LINK_PATTERN = Pattern
      .compile("<([^>]*)>[\\s]*;[\\s]*rel=\"([a-zA-Z0-9]+)\"");
  private static final Pattern PAGE_PATTERN = Pattern.compile("page=(\\d)+");
  private static final String NEXT_LINK = "next";
  public final int code;
  @Nullable
  public final T body;
  @Nullable
  public final String errorMessage;
  @NonNull
  public final Map<String, String> links;

  public ApiResponse(Throwable error) {
    code = 500;
    body = null;
    errorMessage = error.getMessage();
    links = Collections.emptyMap();
  }

  public ApiResponse(Response<T> response) {
    code = response.code();
    if (response.isSuccessful()) {
      body = response.body();
      errorMessage = null;
    } else {
      String message = null;
      if (response.errorBody() != null) {
        try {
          message = response.errorBody().string();
        } catch (IOException ignored) {

        }
      }
      if (message == null || message.trim().length() == 0) {
        message = response.message();
      }
      errorMessage = message;
      body = null;
    }
    String linkHeader = response.headers().get("link");
    if (linkHeader == null) {
      links = Collections.emptyMap();
    } else {
      links = new ArrayMap<>();
      Matcher matcher = LINK_PATTERN.matcher(linkHeader);

      while (matcher.find()) {
        int count = matcher.groupCount();
        if (count == 2) {
          links.put(matcher.group(2), matcher.group(1));
        }
      }
    }
  }

  public boolean isSuccessful() {
    return code >= 200 && code < 300;
  }

  public Integer getNextPage() {
    String next = links.get(NEXT_LINK);
    if (next == null) {
      return null;
    }
    Matcher matcher = PAGE_PATTERN.matcher(next);
    if (!matcher.find() || matcher.groupCount() != 1) {
      return null;
    }
    try {
      return Integer.parseInt(matcher.group(1));
    } catch (NumberFormatException ex) {

      return null;
    }
  }
}

File: app/src/main/java/com/cn29/aac/datasource/api/ApiResponse.kt
package com.cn29.aac.datasource.api

import androidx.collection.ArrayMap
import retrofit2.Response
import java.io.IOException
import java.util.regex.Pattern

class ApiResponse<T> {
    val code: Int
    @JvmField
    val body: T?
    @JvmField
    val errorMessage: String?
    val links: Map<String, String>

    constructor(error: Throwable) {
        code = 500
        body = null
        errorMessage = error.message
        links = emptyMap()
    }

    constructor(response: Response<T>) {
        code = response.code()
        if (response.isSuccessful) {
            body = response.body()
            errorMessage = null
        } else {
            var message: String? = null
            if (response.errorBody() != null) {
                try {
                    message = response.errorBody()!!.string()
                } catch (ignored: IOException) {
                }
            }
            if (message == null || message.trim { it <= ' ' }.length == 0) {
                message = response.message()
            }
            errorMessage = message
            body = null
        }
        val linkHeader = response.headers()["link"]
        if (linkHeader == null) {
            links = emptyMap()
        } else {
            links = ArrayMap()
            val matcher = LINK_PATTERN.matcher(linkHeader)
            while (matcher.find()) {
                val count = matcher.groupCount()
                if (count == 2) {
                    links.put(matcher.group(2), matcher.group(1))
                }
            }
        }
    }

    val isSuccessful: Boolean
        get() = code >= 200 && code < 300

    val nextPage: Int?
        get() {
            val next = links[NEXT_LINK] ?: return null
            val matcher = PAGE_PATTERN.matcher(next)
            return if (!matcher.find() || matcher.groupCount() != 1) {
                null
            } else try {
                matcher.group(1)?.toInt()
            } catch (ex: NumberFormatException) {
                null
            }
        }

    companion object {
        private val LINK_PATTERN = Pattern
                .compile("<([^>]*)>[\\s]*;[\\s]*rel=\"([a-zA-Z0-9]+)\"")
        private val PAGE_PATTERN = Pattern.compile("page=(\\d)+")
        private const val NEXT_LINK = "next"
    }
}